account-compression/sdk/tests/accountCompression.test.ts

```diff
@@ -0,0 +1,684 @@
+      const onChainRoot = getCMTCurrentRoot(splCMT);
+
+      assert(
+        Buffer.from(onChainRoot).equals(offChainTree.root),
+        "Updated on chain root does not match root of updated off chain tree"
+      );
+    });
+  });
+
+  describe(`Having created a tree with depth 3`, () => {
+    const DEPTH = 3;
+    beforeEach(async () => {
+      [splCMTKeypair, offChainTree] = await createTreeOnChain(
+        payer,
+        0,
+        DEPTH,
+        2 ** DEPTH
+      );
+
+      for (let i = 0; i < 2 ** DEPTH; i++) {
+        const newLeaf = Array.from(Buffer.alloc(32, i + 1));
+        const appendIx = createAppendIx(
+          newLeaf,
+          payer,
+          splCMTKeypair.publicKey
+        );
+        await execute(provider, [appendIx], [payer]);
+      }
+
+      // Compare on-chain & off-chain roots
+      const splCMT = await getConcurrentMerkleTree(connection, splCMTKeypair.publicKey);
+
+      assert(
+        getCMTBufferSize(splCMT) === 2 ** DEPTH,
+        "Not all changes were processed"
+      );
+      assert(
+        getCMTActiveIndex(splCMT) === 0,
+        "Not all changes were processed"
+      );
+    });
+
+    it("Random attacker fails to fake the existence of a leaf by autocompleting proof", async () => {
+      const maliciousLeafHash = crypto.randomBytes(32);
+      const maliciousLeafHash1 = crypto.randomBytes(32);
+      const nodeProof: Buffer[] = [];
+      for (let i = 0; i < DEPTH; i++) {
+        nodeProof.push(Buffer.alloc(32));
+      }
+
+      // Root - make this nonsense so it won't match what's in ChangeLog, thus forcing proof autocompletion
+      const replaceIx = createReplaceIx(
+        payer,
+        splCMTKeypair.publicKey,
+        Buffer.alloc(32),
+        maliciousLeafHash,
+        maliciousLeafHash1,
+        0,
+        nodeProof
+      );
+
+      try {
+        await execute(provider, [replaceIx], [payer]);
+        assert(
+          false,
+          "Attacker was able to succesfully write fake existence of a leaf"
+        );
+      } catch (e) { }
+
+      const splCMT = await getConcurrentMerkleTree(connection, splCMTKeypair.publicKey);
+
+      assert(
+        getCMTActiveIndex(splCMT) === 0,
+        "CMT updated its active index after attacker's transaction, when it shouldn't have done anything"
+      );
+    });
+    it("Random attacker fails to fake the existence of a leaf by autocompleting proof", async () => {
+      const maliciousLeafHash = crypto.randomBytes(32);
+      const maliciousLeafHash1 = crypto.randomBytes(32);
+      const nodeProof: Buffer[] = [];
+      for (let i = 0; i < DEPTH; i++) {
+        nodeProof.push(Buffer.alloc(32));
+      }
+
+      // Root - make this nonsense so it won't match what's in CL, and force proof autocompletion
+      const replaceIx = createReplaceIx(
+        payer,
+        splCMTKeypair.publicKey,
+        Buffer.alloc(32),
+        maliciousLeafHash,
+        maliciousLeafHash1,
```

Comment:
Again, to check my understanding, is there anything malicious about this leaf hash? Or is it just a random leaf hash that shouldn't work because the previous one `maliciousLeafHash` is the actual incorrect one

Reply:
Added some docs to explain. 

As an attacker, we want to set `maliciousLeafHash1` by providing `maliciousLeafHash` and `nodeProof` which hash to the current merkle tree root. If we can do this, then we can set leaves to arbitrary values.